File: unique-binary-search-trees.java
class Solution {
    public int numTrees(int n) {
        if (n < 2) {
            return 1;
        }
        int[] dp = new int[n+1];
        dp[0] = 1;
        dp[1] = 1;
        
        for (int i = 2; i <= n; i++) {
            for (int head = 1; head <= i; head++) {
                dp[i] += dp[head-1] * dp[i-head];
            }
        }
        return dp[n];
    }
}

File: unique-binary-search-trees.py
class Solution:
    def numTrees(self, n: int) -> int:
        if n < 2:
            return n
        
        dp = [1, 1] + [0] * (n-1)
        for i in range(2, n+1):
            for root in range(1, i+1):
                left = dp[root - 1]
                right = dp[i-root]
                dp[i] += left * right
        
        return dp[-1]
